web/src/main/kotlin/com/example/ajaxproject/service/GroupChatServiceImpl.kt

```diff
@@ -30,104 +32,89 @@ class GroupChatServiceImpl (
+            .doOnSuccess { _ -> logger.info("Retrieved chat members for chat ID: $chatId") }
     }
 
-    override fun addUserToChat(chatId: String, userId: String): List<User> {
-
-        val chat = groupChatRoomRepository.findChatRoom(chatId)
-        val user = userService.getUserById(userId)
 
-        val existingUser = chat.chatMembers.find { it.id == userId }
-
-        if (existingUser != null) {
-            logger.info("User with ID {} is already present in chat ID {}", userId, chatId)
-            throw WrongActionException("User is already present in this chat")
+    override fun addUserToChat(chatId: String, userId: String): Mono<User> {
+        return Mono.zip(
+            groupChatRoomRepository.findChatRoom(chatId),
+            userService.getById(userId)
+        ).flatMap {
+            val chat = it.t1
+            val user = it.t2
+
+            val existingUser = chat.chatMembers.find { it.id == userId }
+            if (existingUser != null) {
+                logger.info("User with ID $userId is already present in chat ID $chatId")
+                Mono.just(existingUser)
+            } else {
+                chat.chatMembers += user
+                logger.info("User with ID $userId added to chat ID $chatId")
+                groupChatRoomRepository.save(chat).map { user }
+            }
         }
-
-        chat.chatMembers += user
-        logger.info("User with ID {} added to chat ID {}", userId, chatId)
-
-        val updatedChatRoom = groupChatRoomRepository.save(chat)
-
-        return updatedChatRoom.chatMembers
-
     }
 
     @Notification
-    override fun sendMessageToGroup(groupChatDto: GroupChatDto): GroupChatMessageResponse {
-
+    override fun sendMessageToGroup(groupChatDto: GroupChatDto): Mono<GroupChatMessageResponse> {
         val chatMessage = GroupChatMessage(
             id = ObjectId().toHexString(),
             groupChatRoom = groupChatDto.chatId,
             senderId = groupChatDto.senderId,
             message = groupChatDto.message,
             date = Date()
         )
-
-        logger.info("Message sent to group chat ID: {}" , groupChatDto.chatId)
-
-        groupChatMessageRepository.save(chatMessage)
-
-        return toResponseDto(chatMessage)
-
+        return groupChatMessageRepository.save(chatMessage)
+            .doOnSuccess { logger.info("Message sent to group chat ID: ${groupChatDto.chatId}") }
+            .map { toResponseDto(it) }
     }
 
-    override fun getAllGroupMessages(chatId: String): List<GroupChatMessageResponse> {
-
-        val messages = groupChatMessageRepository.findAllMessagesInChat(chatId)
-
-        return messages.map { toResponseDto(it) }
+    @Cacheable("getAllGroupMessages")
+    override fun getAllGroupMessages(chatId: String): Mono<List<GroupChatMessageResponse>> {
+        return groupChatMessageRepository.findAllMessagesInChat(chatId)
+            .map { toResponseDto(it) }
+            .collectList()
+            .cache()
```

Comment:
```suggestion
```
Why do you want caching it?

Reply:
You was right - its enough to have only @Cacheable